File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/persistence/MemoryPersistenceImplTest.java
package com.godaddy.asherah.appencryption.persistence;

import static org.junit.jupiter.api.Assertions.*;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class MemoryPersistenceImplTest {
  MemoryPersistenceImpl<String> memoryPersistenceImpl;

  @BeforeEach
  void setUp() {
    memoryPersistenceImpl = new MemoryPersistenceImpl<>();
  }

  @Test
  void testLoadAndStoreWithValidKey() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "This is my value";

    memoryPersistenceImpl.store(keyId, created, value);

    String retrievedValue = memoryPersistenceImpl.load(keyId, created).get();

    assertEquals(value, retrievedValue, "Value loaded from persistence doesn't match what was stored.");
  }

  @Test
  void testLoadAndStoreWithInvalidKey() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "This is my value";

    memoryPersistenceImpl.store(keyId, created, value);

    assertFalse(memoryPersistenceImpl.load("some non-existent key", created).isPresent(),
        "Found value for non-existent key");
  }

  @Test
  void testLoadLatestValueMultipleCreatedAndValuesForKeyIdShouldReturnLatest() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "created value";
    memoryPersistenceImpl.store(keyId, created, value);

    Instant createdOneHourLater = created.plus(1, ChronoUnit.HOURS);
    String valueCreatedOneHourLater = value + createdOneHourLater;
    memoryPersistenceImpl.store(keyId, createdOneHourLater, valueCreatedOneHourLater);

    Instant createdOneDayLater = created.plus(1, ChronoUnit.DAYS);
    String valueCreatedOneDayLater = value + createdOneDayLater;
    memoryPersistenceImpl.store(keyId, createdOneDayLater, valueCreatedOneDayLater);

    Instant createdOneWeekEarlier = created.minus(7, ChronoUnit.DAYS);
    String valueCreatedOneWeekEarlier = value + createdOneWeekEarlier;
    memoryPersistenceImpl.store(keyId, createdOneWeekEarlier, valueCreatedOneWeekEarlier);

    assertEquals(valueCreatedOneDayLater, memoryPersistenceImpl.loadLatestValue(keyId).get());
  }

  @Test
  void testLoadLatestValueNonExistantKeyIdShouldReturnNull() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "created value";
    memoryPersistenceImpl.store(keyId, created, value);

    assertFalse(memoryPersistenceImpl.loadLatestValue("some non-existent key").isPresent());
  }

  @Test
  void testStoreWithDuplicateKeyShouldReturnFalse() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "this is the value";

    assertTrue(memoryPersistenceImpl.store(keyId, created, value));
    assertFalse(memoryPersistenceImpl.store(keyId, created, value));
  }
}

File: languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Persistence/MemoryPersistenceImplTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Persistence;
using LanguageExt;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Persistence
{
    [Collection("Logger Fixture collection")]
    public class MemoryPersistenceImplTest
    {
        private readonly MemoryPersistenceImpl<string> memoryPersistenceImpl;

        public MemoryPersistenceImplTest()
        {
            memoryPersistenceImpl = new MemoryPersistenceImpl<string>();
        }

        [Fact]
        private void TestLoadAndStoreWithValidKey()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            memoryPersistenceImpl.Store(keyId, created, value);

            Option<string> actualValue = memoryPersistenceImpl.Load(keyId, created);

            Assert.Equal(value, actualValue);
        }

        [Fact]
        private void TestLoadAndStoreWithInvalidKey()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            memoryPersistenceImpl.Store(keyId, created, value);

            Option<string> actualValue = memoryPersistenceImpl.Load("some non-existent key", created);

            Assert.True(actualValue.IsNone);
        }

        [Fact]
        private void TestLoadLatestValueMultipleCreatedAndValuesForKeyIdShouldReturnLatest()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            memoryPersistenceImpl.Store(keyId, created, value);

            DateTimeOffset createdOneHourLater = created.AddHours(1);
            string valueCreatedOneHourLater = value + createdOneHourLater;
            memoryPersistenceImpl.Store(keyId, createdOneHourLater, valueCreatedOneHourLater);

            DateTimeOffset createdOneDayLater = created.AddDays(1);
            string valueCreatedOneDayLater = value + createdOneDayLater;
            memoryPersistenceImpl.Store(keyId, createdOneDayLater, valueCreatedOneDayLater);

            DateTimeOffset createdOneWeekEarlier = created.AddDays(-7);
            string valueCreatedOneWeekEarlier = value + createdOneWeekEarlier;
            memoryPersistenceImpl.Store(keyId, createdOneWeekEarlier, valueCreatedOneWeekEarlier);

            Option<string> loadLatestValue = memoryPersistenceImpl.LoadLatestValue(keyId);

            Assert.Equal(valueCreatedOneDayLater, loadLatestValue);
        }

        [Fact]
        private void TestLoadLatestValueNonExistentKeyIdShouldReturnNull()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            memoryPersistenceImpl.Store(keyId, created, value);

            Assert.True(memoryPersistenceImpl.LoadLatestValue("some non-existent key").IsNone);
        }

        [Fact]
        private void TestStoreWithDuplicateKeyShouldReturnFalse()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            Assert.True(memoryPersistenceImpl.Store(keyId, created, value));
            Assert.False(memoryPersistenceImpl.Store(keyId, created, value));
        }
    }
}
